docs/reference/transforms/matrix.rst

```diff
@@ -0,0 +1,210 @@
+.. _matrix transforms:
+
+Matrix Transforms
+=================
+
+The :mod:`taskgraph.transforms.matrix` transforms can be used to split a base
+task into many subtasks based on a defined matrix.
+
+These transforms are useful if you need to have many tasks that are very
+similar except for some small configuration differences.
+
+Usage
+-----
+
+Add the transform to the ``transforms`` key in your ``kind.yml`` file:
+
+.. code-block:: yaml
+
+   transforms:
+     - taskgraph.transforms.matrix
+     # ...
+
+Then create a ``matrix`` section in your task definition, e.g:
+
+.. code-block:: yaml
+
+   tasks:
+     test:
+       matrix:
+         os: ["win", "mac", "linux"]
+
+       # rest of task definition
+
+This will split the ``test`` task into three; ``test-win``, ``test-mac`` and
+``test-linux``.
+
+Matrix with Multiple Rows
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can add as many rows as you like to the matrix, and every combination of
+tasks will be generated. For example, the following matrix:
+
+.. code-block:: yaml
+
+   tasks:
+     test:
+       matrix:
+         os: ["win", "mac", "linux"]
+         python: ["py312", "py311"]
+
+       # rest of task definition
+
+Will generate these tasks:
+
+- ``test-win-py312``
+- ``test-win-py311``
+- ``test-mac-py312``
+- ``test-mac-py311``
+- ``test-linux-py312``
+- ``test-linux-py311``
+
+Note that the name of the tasks will be built based on the order of rows in the
+matrix.
+
+Substituting Matrix Context into the Task Definition
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Of course these tasks will be identical, so you'll want to change other parts
+of the task definition based on the matrix values.
+
+Substituting Values in Yaml
+```````````````````````````
+
+The simplest way to change a matrix task's definition, is to use the built-in
+yaml substitution:
+
+.. code-block:: yaml
+
+   tasks:
+     test:
+       matrix:
+         os: ["win", "mac", "linux"]
+       description: Run {matrix[os]} tests
+       worker-type: {matrix[os]}-worker
+
+Limiting Substitution
+'''''''''''''''''''''
+
+By default, all keys and values in the task definition will be checked for
+substitution parameters. But in some cases, it might be desirable to limit which
+keys get substituted, such as when using the ``matrix`` transforms alongside
+other transforms that perform substitution, such as the
+:mod:`~taskgraph.transforms.task_context` or
+:mod:`~taskgraph.transforms.chunking` transforms.
+
```

Comment:
In translations, we've had cases where limiting to a particular field is not enough because we need to format the same value in multiple transforms. I ended up implementing a [partial substitution function](https://github.com/mozilla/firefox-translations-training/blob/d1d1efc441bfa9fc0f3e2c176f778197d16899df/taskcluster/translations_taskgraph/util/substitution.py) to handle these.

I'm not super eager to have that upstreamed, but as we add more and more very general transforms it may become necessary.

(The `task-context` part below may help reduce the need for this, but it can still come up elsewhere. Eg: in translations we have places where we have chunk numbers being substituted as well as other things.)

Reply:
Yeah, I realized we have a lot of transforms that do some kind of substitution now..

By my count there's:

* task_context
* chunking
* from_deps
* run (substitutes `cwd: {checkout}`)
* and now matrix

Perhaps we need to consolidate this somehow. Maybe each one of these transforms could build up `**subs` in an attribute, then the `task` transforms could do a single substitution at the very end? I'm not sure if that would work for all cases or not though.

Reply:
Yeah, that seems plausible for many cases. If, eg: we ever have a transform that cares about a value that an earlier transform may substitute it could be annoying. I don't think any of this needs to block this work though, for the record.